Agenda for 2024-10-09 CG meeting (#688)
solid/specification: meetings/2024-10-09.md

### Scribes
* [elf Pavlik](https://elf-pavlik.hackers4peace.net)

### Introductions
* Niko: Developer of NextGraph, based on semantic web. Local-first and end-to-end encrypted.
* Damon: Co founder of WeDistribute, indymedia org, covering decentralized social tech. Lead project manager behind memory project, building on top of ActivityPods.

---

## Topics

### NextGraph overview

URL: https://docs.nextgraph.org/en/introduction/

* eP: Niko who is developing NextGraph will give 15-20min presentation
* Niko: presenting slides and demo

> * ...: Peer to Peer semantic we collaboration

Reviewer: ```suggestion
* ...: Peer to Peer semantic web collaboration
```

* ...: NextGraph is doing the same as Solid but differently
* ...: RDF, Linked Data, SPARQL

> * ...: Local-first CRDT of RDF - you can edit offline and data gets synchronized once you get online and without conflicts. I developed CRDTs for RDF, there are already formats for JSON. I added compatibility with automerge and yjs. This way one can mix all kinds of data. 
> * ...: end-to-end encrypted
> * ...: decentralized - 2 tier. Not peer-to-peer we have servers which act as brokers. It is based on pub/sub where replicas subscribe. All the changes are propageted. Brokers are constantly exchanging updates in anticipation for the devices that subscribe to them. Local replica offers SPARQL locally to the applictions.

Reviewer: ```suggestion
* ...: Local-first CRDT of RDF — you can edit offline and data gets synchronized once you get online and without conflicts. I developed CRDTs for RDF, there are already formats for JSON. I added compatibility with automerge and yjs. This way one can mix all kinds of data. 
* ...: end-to-end encrypted
* ...: decentralized — 2 tier. Not peer-to-peer; we have servers which act as brokers. It is based on pub/sub where replicas subscribe. All the changes are propageted. Brokers are constantly exchanging updates in anticipation of the devices that subscribe to them. Local replica offers SPARQL locally to the applications.
```

* ...: also binary data

> * ...: we have an app mobile, desktopand web - alpha

Reviewer: ```suggestion
* ...: we have an app mobile, desktop and web — alpha
```

* ...: a framework soon, Rust/JS&TS/Svelte
* ...: a new protocol (binary) not using HTTP, no domain names and certs, also for efficiency

> * ...: we are collaborating with ActivityPods, already mentioned today, yesterday we annonced the collabortion. On Monday activitypods released v2.0 - it combines ActivityPub with Solid.
> * ...: ActivityPods also uses SPARQL, solid documents are named graphs. Next Graph is doing the same which fits well. They will also adopt DIDs for documents and agents, as NextGraph. A special DID method will be registered.

Reviewer: ```suggestion
* ...: we are collaborating with ActivityPods, already mentioned today, yesterday we announced the collaboration. On Monday, activitypods released v2.0. It combines ActivityPub with Solid.
* ...: ActivityPods also uses SPARQL; Solid documents are named graphs. NextGraph is doing the same which fits well. They will also adopt DIDs for documents and agents, as NextGraph. A special DID method will be registered.
```

> * ...: Activity Pods are using SAI - Solid Application Interop, NextGraph is also using it.
> * ...: long term goal is unify our frameworks

Reviewer: ```suggestion
* ...: Activity Pods are using SAI — Solid Application Interop. NextGraph is also using it.
* ...: long term goal is to unify our frameworks
```

* ...: reactive stores in React, Svelte, Vue
* ...: NextGraph has documents which are RDF resources, which have primary class.

> * ...: viewers / editors are the apps, manifest specifies which primary classes it can read and write
> * ...: end user can choose which viewer / editor it uses, user is the owenr of the data

Reviewer: ```suggestion
* ...: viewers and editors are the apps; manifest specifies which primary classes each can read and write.
* ...: end user can choose which viewer and/or editor it uses; user is the owner of the data.
```

> * ...: plans for everything app, to give one good app, there can be other apps available but we want to have one good app on every domain.
> * ...: it is also a social network thanks to pub/sub, it can be compatible with activitypub, also can be bridged to other protocols
> * ...: data is based on revisions, each change is a commit, one can come back to any past revision

Reviewer: ```suggestion
* ...: plans for "everything" app, to give one good app; there can be other apps available, but we want to have one good app on every domain.
* ...: it is also a social network. Thanks to pub/sub, it can be compatible with activitypub, and can also be bridged to other protocols.
* ...: data is based on revisions. Each change is a commit; one can come back to any past revision.
```

> * ...: content is signed by the author, also there are group signatures to proof authenticity to the outsiders
> * ...: it also has branches, new branches can be created as blocks, document can be composed of blocks
> * ...: we also have consensus based on threshold signatures, similar to how blockchains work

Reviewer: ```suggestion
* ...: content is signed by the author. There are also group signatures to prove authenticity to outsiders.
* ...: it also has branches. New branches can be created as blocks; document can be composed of blocks.
* ...: we also have consensus based on threshold signatures, similar to how blockchains work.
```

* ...: synchronous transactions (ACID)
* ...: thanks to NLnet for supporting the project
* ...: engaging in talks with other actors to converge and gather forces to defeat Big Tech
* ...: demo!

> * ...: we have web version on left and native linux on the right. there are three stores, one is private which user can only share across devices. there is also the shared space, where one can interact with other users. There is also a public space, like a website.
> * ...: I will create a new document, we can select from different primary classes. As you see it synchronized. I can change the title, it also updated in the other app. I can do a SPARQL update.

Reviewer: ```suggestion
* ...: we have web version on the left and native linux on the right. There are three stores: one is private, which only the user can share across devices. There is also the shared space, where one can interact with other users. There is also a public space, like a website.
* ...: I will create a new document. We can select from different primary classes. As you can see, it synchronized. I can change the title; it is also updated in the other app. I can do a SPARQL update.
```

* ...: you can see `did:ng:o:.....` the last part has a public key. 

> * ...: let's do another update, now in the other app we see the whole history of the document, creation, signature, adding first triple, then adding another triple. I will add another one on the left and we can see another commit on the right. 
> * ...: Now I will disable connection in the web app, I can still edit the document. And in the other ap I will make a different update. No sync yet since the webapp is offline. Now we go backonline and we see on the right a fork. As soon as I make another commit we can see the automatic merge. So it has the automerge without conflicts, it also work on data deletion.

Reviewer: ```suggestion
* ...: Let's do another update. Now, in the other app, we see the whole history of the document: creation, signature, adding first triple, then adding another triple. I will add another one on the left, and we can see another commit on the right. 
* ...: Now I will disable connection in the web app, I can still edit the document. And in the other app, I will make a different update. No sync yet since the web app is offline. Now we go back online, and we see on the right a fork. As soon as I make another commit, we can see the automatic merge. So it has the automerge without conflicts. It also works on data deletion.
```

* Rui: what does the `o` mean in the DID identifier?

> * Niko: We are registering the `did:ng` method, the rest is specific to the next graph and `o` means a document. You can find it in the DID section of NextGraph docs

Reviewer: ```suggestion
* Niko: We are registering the `did:ng` method; the rest is specific to NextGraph, and `o` means a document. You can find it in the DID section of NextGraph docs.
```

> * Rahul: Are you aware of Georges m-ld? How does it compare. Also what transport are you using? Are you aware of CRDTs algorithms developed by braid?
> * Niko: Good remarks, I'm familiar with m-ld. I still need to talk with George. m-ld is not end-to-end encrypted but we need to discuss details. We use websockets for now on every clients, also between brokers we use websockets for practical reasons. As for braid, I met them a week ago, I will join them again in the week. A lot of common grounds and interest to work together. 
> * Tim: Does it use UDP or TCP
> * Niko: it uses WebSockets so TCP. For now it's good.

Reviewer: ```suggestion
* Rahul: Are you aware of Georges `m-ld`? How does it compare? Also, what transport are you using? Are you aware of CRDTs algorithms developed by braid?
* Niko: Good remarks, I'm familiar with `m-ld`. I still need to talk with George. `m-ld` is not end-to-end encrypted, but we need to discuss details. We use websockets for now on every client; we also use websockets between brokers, for practical reasons. As for braid, I met them a week ago; I will join them again in a week. A lot of common ground and interest to work together. 
* Tim: Does it use UDP or TCP?
* Niko: it uses WebSockets, so TCP. For now it's good.
```

> * Tim: This is very exciting, it has a lot of things we want in Solid. We want local-first. Where are you based and who do you work for, why did you start NextCloud.
> * Niko: I'm independent, sponsored by NLnet using EU grants last two years. I applied for another grant this year. I really like your vision about the giant global graph, very inspired by that. I worked with property graph first, four years I embraced RDF. Later added CRDTs and end-to-end encryption. Trying to bridge a gap with solid.
> * Hadrian: Is new method neccessary? There is `did:key`, you may have problem with key rotation.

Reviewer: ```suggestion
* Tim: This is very exciting; it has a lot of things we want in Solid. We want local-first. Where are you based, and who do you work for? Why did you start NextCloud?
* Niko: I'm independent, sponsored by NLnet using EU grants last two years. I applied for another grant this year. I really like your vision about the giant global graph, very inspired by that. I worked with property graphs first; four years ago, I embraced RDF. Later added CRDTs and end-to-end encryption. Trying to bridge a gap with Solid.
* Hadrian: Is new method necessary? There is `did:key`; you may have problems with key rotation.
```

> * Niko: yes key rotation is very important, it is already integrated into protocol and the design. The public key is never going to change, but it is not used for signing. There is a chain, this can be revoked and updated.
> * Hadrian: a discussion for another time
> * Niko: please feel invited to nextgraph forum

Reviewer: ```suggestion
* Niko: yes, key rotation is very important. It is already integrated into the protocol and the design. The public key is never going to change, but it is not used for signing. There is a chain; this can be revoked and updated.
* Hadrian: A discussion for another time.
* Niko: Please feel invited to NextGraph forum.
```

### Solid Practicioners

URL: https://github.com/solid-contrib/practitioners?tab=readme-ov-file#communication-channels

* eP: I would like to propose adding the bi-weekly meeting to the CG calendar 
* ep: Practitioners are meeting every other Thursday. Proposal to move it to the CG. The Solid CG calendar would be a good place.
* Hadrian: no objection from me, this is related to the direction of Solid CG, are Practicioners open to that?
* Rahul: Jeff wanted CG involved from the very beggining, adding to the CG calendar would be the minimum involvement.
* Hadrian: I would prefer not to make that decision until we get agreement with two other chairs.
* Rahul: Can we coordiante it over matrix?

ACTION: Hadrian to coordinate with other chairs and then Jeff

### Special Topic Meeting scheduling 

* CRDT
* E2EE
* eP: on Tuesday I have conflict with FedCM weekly. We should do an online scheduling poll template and use it to schedule each STM

> * eP: Solid calendar app and integrated schedule polling can be a **use case**. With conflicts detector and templates for starting a poll and responding.
> * Hadrian: I have a conflict as well. One proposal was to have two times for different timezones.
> * eP: Propose to remove the one that's in the calendar and we can do online polls and then we make sure that the experts in the field can be present. Better than one fixed slot. For instance we can run polls for 2 weeks ahead and reach out to the expert.

Reviewer: ```suggestion
* eP: Solid calendar app and integrated schedule polling can be a **use case**. With conflict detector, and templates for starting and responding to a poll.
* Hadrian: I have a conflict as well. One proposal was to have two times, for different timezones.
* eP: Propose to remove the one that's in the calendar. We can do online polls and make sure that the experts in the field can be present. Better than one fixed slot. For instance, we can run polls 2 weeks ahead, and reach out to the experts.
```

> * Tim: A year or two ago, we had SolidOS meeting and decided to have them all in the same constant GMT timezone. We should keep it fixet to the GMT.
> * Rahul: I see this coordination problem, especially when you want to get the timezones. I would still suggest to find antohter fixed slot, and keep it reserved for STM. Having it there would make things easier.
> * Tim: You could cooridnate with different priority on different weeks.

Reviewer: ```suggestion
* Tim: A year or two ago, we had SolidOS meetings, and decided to have them all in the same constant GMT timezone. We should keep it fixed to GMT.
* Rahul: I see this coordination problem, especially when you want to get the timezones. I would still suggest to find another fixed slot, and keep it reserved for STM. Having it there would make things easier.
* Tim: You could coordinate with different priority on different weeks.
```

> * eP: We had this slot and didn't really use it. It clearly doesn't work. For me it conflicts. Suggest we start with online slots. After a few months we can decide if to use the same slot. If all experts are using the same slot then we can continue that way.
> * Hadrian: we have proposal for a poll and for fixed slots
> * eP: I don't think the two proposals are mutually exclusive. Especially if we want to reach out to experts with are not Solid CG regulars. 
> * Hadrain: I'm propose reducing STM meeting to one hour, have multiple slots in the week which would be empty. Everyone can become champion of the meeting and can reserve slot and ensure experts are present.

Reviewer: ```suggestion
* eP: We had this slot and didn't really use it. It clearly doesn't work. For me, it conflicts. Suggest we start with online slots. After a few months, we can decide whether to use the same slot. If all experts are using the same slot, we can continue that way.
* Hadrian: We have proposal for a poll and for fixed slots.
* eP: I don't think the two proposals are mutually exclusive. Especially if we want to reach out to experts who are not Solid CG regulars. 
* Hadrain: I'm propose reducing STM meetings to one hour, and having multiple slots in the week which would be empty to start. Anyone can become a champion of a meeting, reserve a slot, and ensure experts are present.
```

* eP: We need to do the poll first.
* Hadrian: You can do it in any order you like. 

ACTION: Pavlik to create a poll for CRDTs

### Use cases (continue from last week)
 * eP: eating one's own cat food ☝️ 
 * eP: https://github.com/janeirodigital/sai-js/blob/main/examples%2Fplenary%2FREADME.md
 * eP: example apps that could be adapted to solid
   * https://www.discourse.org/ (forum)
   * https://about.gitlab.com/ (project management)
   * https://huly.app (project management)

>    * https://openfoodnetwork.org/software-platform/ (supply chains - food)

Reviewer: ```suggestion
   * https://openfoodnetwork.org/software-platform/ (supply chains — food)
```
